sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteArithmeticFiltersOnIntOrLongColumn.scala

```diff
@@ -0,0 +1,147 @@
+import org.apache.spark.sql.types.{DataType, IntegerType, LongType}
+
+/**
+ * Rewrite arithmetic filters on int or long column to its equivalent form,
+ * leaving attribute alone in one side, so that we can push it down to
+ * parquet or other file format.
+ * For example, this rule can optimize
+ * {{{
+ *   SELECT * FROM table WHERE i + 3 = 5
+ * }}}
+ * to
+ * {{{
+ *   SELECT * FROM table WHERE i = 5 - 3
+ * }}}
+ * when i is Int or Long, and then other rules will further optimize it to
+ * {{{
+ *   SELECT * FROM table WHERE i = 2
+ * }}}
+ * The arithmetic operation supports `Add` and `Subtract`. The comparision supports
+ * '=', '>=', '<=', '>', '<', '!='. It only supports type of `INT` and `LONG`,
+ * it doesn't support `FLOAT` or `DOUBLE` for precision issues.
+ */
+object RewriteArithmeticFiltersOnIntOrLongColumn extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case f: Filter =>
+      f transformExpressionsUp {
+        case e @ BinaryComparison(left: BinaryArithmetic, right: Expression)
+            if right.foldable && isDataTypeSafe(left.dataType) =>
+          transformLeft(e, left, right)
+        case e @ BinaryComparison(left: Expression, right: BinaryArithmetic)
+            if left.foldable && isDataTypeSafe(right.dataType) =>
+          transformRight(e, left, right)
+      }
+  }
+
+  private def transformLeft(
+      bc: BinaryComparison,
+      left: BinaryArithmetic,
+      right: Expression): Expression = {
+    left match {
+      case Add(ar: AttributeReference, e) if e.foldable && isOptSafe(Subtract(right, e)) =>
+        bc.makeCopy(Array(ar, Subtract(right, e)))
+      case Add(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(right, e)) =>
+        bc.makeCopy(Array(ar, Subtract(right, e)))
+      case Subtract(ar: AttributeReference, e) if e.foldable && isOptSafe(Add(right, e)) =>
+        bc.makeCopy(Array(ar, Add(right, e)))
+      case Subtract(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(e, right)) =>
+        bc.makeCopy(Array(Subtract(e, right), ar))
+      case _ => bc
+    }
+  }
+
+  private def transformRight(
+      bc: BinaryComparison,
+      left: Expression,
+      right: BinaryArithmetic): Expression = {
+    right match {
+      case Add(ar: AttributeReference, e) if e.foldable && isOptSafe(Subtract(left, e)) =>
+        bc.makeCopy(Array(Subtract(left, e), ar))
+      case Add(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(left, e)) =>
+        bc.makeCopy(Array(Subtract(left, e), ar))
+      case Subtract(ar: AttributeReference, e) if e.foldable && isOptSafe(Add(left, e)) =>
+        bc.makeCopy(Array(Add(left, e), ar))
+      case Subtract(e, ar: AttributeReference) if e.foldable && isOptSafe(Subtract(e, left)) =>
+        bc.makeCopy(Array(ar, Subtract(e, left)))
+      case _ => bc
+    }
+  }
+
+  private def isDataTypeSafe(dataType: DataType): Boolean = dataType match {
```

Comment:
why only integer and longs are accepted?

Reply:
`Float` and `Double` has precision issues. For example, `a + 3.2 = 4.0` will be convert to `a  < 0.7999999999999998`.

Reply:
How about the other integral types, e.g., short?

Reply:
yes, integral types(byte, short, int, long) are all ok. I'll add support for byte and short type as well.

Reply:
I think the precision issue would be there anyway, when executed at runtime, am I wrong?

Reply:
> I think the precision issue would be there anyway, when executed at runtime, am I wrong?

I have a simple test on a table with a Double type column `a`, it has two records: 0.7999999999999998 and 0.8
![image](https://user-images.githubusercontent.com/1312321/54008703-d07ef780-41a2-11e9-96cd-136764253b25.png)

with `a + 3.2 = 4.0`, it returns both two records. But if we optimized it to `a = 0.7999999999999998`, the result will be wrong

Reply:
Here is an example:
```
scala> spark.sql("select float(1E-8) + float(1E+10) <= float(1E+10)").show()
+----------------------------------------------------------------------+
|((CAST(1E-8 AS FLOAT) + CAST(1E+10 AS FLOAT)) <= CAST(1E+10 AS FLOAT))|
+----------------------------------------------------------------------+
|                                                                  true|
+----------------------------------------------------------------------+


scala> spark.sql("select float(1E-8) <= float(1E+10) - float(1E+10)").show()
+----------------------------------------------------------------------+
|(CAST(1E-8 AS FLOAT) <= (CAST(1E+10 AS FLOAT) - CAST(1E+10 AS FLOAT)))|
+----------------------------------------------------------------------+
|                                                                 false|
+----------------------------------------------------------------------+

```
Although `float(1E-8) + float(1E+10) <= float(1E+10)` should return false. This may lead to inconsistency.

Reply:
thanks for the explanation